Use Enterprise Subscription on Enterprise environments

esoergel opened the pull request:
https://dimagi-dev.atlassian.net/browse/SC-64

##### PRODUCT DESCRIPTION
Make all domains on enterprise environments use the enterprise subscription

##### SUMMARY
Enterprise environments don't need subscriptions - we typically enable "Dimagi Only CommCare Enterprise Edition" manually on these environments, but it must be done for every domain, and we already hide links to the subscription pages, so doing so is unintuitive.  We instruct people to enable it for the command line, but it's easy to forget to do this, especially on secondary domains.

This PR proposes an alternative approach - setting `ENTERPRISE` as the default plan, and always returning that.  It looks everything goes through `Subscription.get_active_subscription_by_domain` or `Subscription.get_subscribed_plan_by_domain` (which calls the former), and both use `DefaultProductPlan.get_default_plan_version`, so that's what I modified here.

This feels like a fairly impactful change, so I'd be interested to hear if anyone has a proposal for ensuring that it works as I'd hope - are there weird edge cases where it will select a plan through some other means?  It seems to behave well when I click around locally.

@snopoke @calellowitz pinging you as I see your names in git blame for much of the `ENTERPRISE_MODE` stuff.  I'd initially planned on making a new flag for this purpose, but this seems to be very much in-line with how I'd imagine that flag to work.  Have you considered this sort of approach before?  Any concerns I might not have thought of?

An alternative approach here would be to modify `request_new_domain` to always create an enterprise subscription when new domains are created.  That feels a bit messier (there's no need for them, really - and what if they need to be updated?), though perhaps a bit safer.

##### FEATURE FLAG

esoergel committed:
Add failing test for desired behavior

esoergel committed:
Default to Enterprise in ENTERPRISE_MODE

esoergel committed:
Don't create a subscription on enterprise envs

snopoke commented on corehq/apps/accounting/models.py:
```diff
@@ -787,7 +787,9 @@ class Meta(object):
     @classmethod
     def get_default_plan_version(cls, edition=None, is_trial=False,
```
this might need to be cached now since it will get hit on every request instead of the already cached `get_active_subscription_by_domain`

esoergel replied:
Good point - and it'll be a very small cache too, since there's little variance on these parameters.

snopoke reviewed (commented):
This looks like a good approach to me. I'm not familiar enough with billing to know if there are any edge cases.

esoergel commented:
Flagging also that this will affect ICDS - realize I didn't explicitly state that above.

esoergel committed:
Cache default plan lookup

snopoke commented on corehq/apps/accounting/models.py:
```diff
@@ -785,9 +785,12 @@ class Meta(object):
         unique_together = ('edition', 'is_trial', 'is_report_builder_enabled')
 
     @classmethod
+    @quickcache(['edition', 'is_trial', 'is_report_builder_enabled'], timeout=24 * 60 * 60)
```
If this is cached for 24 hours it will take that long before a new version becomes active. Is that likely to be an issue? I'm not sure how often new default plans are created or quite what the process is.

esoergel replied:
default plans are created only rarely, via a db migration.  I believe this was the most recent change:
https://github.com/dimagi/commcare-hq/blob/0f8ade4bee86bab84fa65c9de1294f200ada0db5/corehq/apps/accounting/migrations/0046_new_plans.py#L26-L27
I'll add a cache clear to that command actually, because it would be confusing for that to take a day to kick in.

snopoke commented on corehq/apps/accounting/models.py:
```diff
@@ -1730,6 +1733,9 @@ def get_active_subscription_by_domain(cls, domain_name_or_obj):
     @classmethod
     @quickcache(['domain_name'], timeout=60 * 60)
```
how about adding this to avoid hitting the cache as well:
```suggestion
    @quickcache(['domain_name'], timeout=60 * 60, skip_arg=lambda *args: settings.ENTERPRISE_MODE)
```

esoergel replied:
makes sense

esoergel committed:
Update corehq/apps/accounting/models.py

Co-Authored-By: Simon Kelly <skelly@dimagi.com>

esoergel committed:
Ultimately futile attempt to invalidate this cache

I wasn't able to get tests to pass with any means of cache invalidation.  Here's
what was attempted.  I'm gonna remove this in the next commit and only cache
this in ENTERPRISE_MODE

esoergel committed:
Give up on trying to use caching in tests

This breaks in all sorts of fun and exciting ways.  Figure I'll enable
the cache only when I know it's needed - in production enterprise
environments

esoergel committed:
Move enterprise mode fork out of cache

esoergel committed:
Clear cache when fiddling with plans

sravfeyn commented on corehq/apps/accounting/models.py:
```diff
@@ -785,9 +785,13 @@ class Meta(object):
 
     @classmethod
+    @quickcache(['edition', 'is_trial', 'is_report_builder_enabled'],
+                skip_arg=lambda *args, **kwargs: not settings.ENTERPRISE_MODE or settings.UNIT_TESTING)
```
Why skip cache for ENTERPRISE envs, line 796 `DefaultProductPlan.objects.select_related` would still be evaluated for enterprise envs, isn't it?

snopoke replied:
I think this says the reverse: `skip if not settings.ENTERPRISE_MODE`. In other words, only cache this for enterprise envs.

sravfeyn replied:
My bad, thanks

esoergel commented:
(this is undergoing QA at the moment)

esoergel committed:
Move similar lines closer to each other

esoergel committed:
Only worry about billing on non-enterprise envs

esoergel committed:
Pull out subscription stuff

esoergel commented:
QA passed, going to merge on Monday so the rollout to ICDS happens on the normal schedule
